src/shape.rs

```diff
@@ -660,237 +677,214 @@ impl ShapeLine {
-                            } else if let Some(previous_word) = span.words.get(i - 1) {
-                                // Current word causing a wrap is not whitespace, so we ignore the
-                                // previous word if it's a whitespace
-                                if previous_word.blank {
-                                    number_of_blanks -= 1;
-                                    prev_word_width = Some(previous_word.x_advance * font_size);
+                            let mut trialing_space_width = None;
+                            if i > 0 {
+                                if let Some(previous_word) = span.words.get(i - 1) {
+                                    // Current word causing a wrap is not whitespace, so we ignore the
+                                    // previous word if it's a whitespace
+                                    if previous_word.blank {
+                                        trialing_space_width =
+                                            Some(previous_word.x_advance * font_size);
+                                        number_of_blanks = number_of_blanks.saturating_sub(1);
+                                    }
                                 }
                             }
-                            if let Some(width) = prev_word_width {
-                                word_ranges.push((
-                                    (i, 0),
+                            if let Some(width) = trialing_space_width {
+                                self.add_to_visual_line(
+                                    &mut current_visual_line,
+                                    span_index,
+                                    (i + 2, 0),
                                     fitting_start,
                                     word_range_width - width,
                                     number_of_blanks,
-                                ));
+                                );
                             } else {
-                                word_ranges.push((
+                                self.add_to_visual_line(
+                                    &mut current_visual_line,
+                                    span_index,
                                     (i + 1, 0),
                                     fitting_start,
                                     word_range_width,
                                     number_of_blanks,
-                                ));
+                                );
                             }
+                            visual_lines.push(current_visual_line);
+                            current_visual_line = VisualLine::default();
+
                             number_of_blanks = 0;
                             if word.blank {
                                 fit_x = line_width;
                                 word_range_width = 0.;
                                 fitting_start = (i + 1, 0);
```

Comment:
I also don't understand why this is `i + 1` instead of `i - 1` if we are going in the opposite direction.